### PR TITLE
[NFC] Reduce SIL's reliance on FunctionType::getInput

### PR DESCRIPTION
**Pull request by CodaFi**

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test and merge

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -2853,16 +2853,33 @@ bool AnyFunctionType::isCanonicalFunctionInputType(Type input) {
   return isa<TypeVariableType>(input.getPointer());
 }
 
+void AnyFunctionType::substParams(const SubstitutionMap &substitutions,
```
Comment (slavapestov) [inline lib/AST/Type.cpp]: I don't think this is necessary

File: lib/SILGen/SILGenType.cpp [hunk 2]
```diff
@@ -568,7 +568,9 @@ SILGenModule::emitProtocolWitness(ProtocolConformance *conformance,
     auto reqtSubs = witness.getRequirementToSyntheticSubs();
     auto reqtSubMap = reqtOrigTy->getGenericSignature()
         ->getSubstitutionMap(reqtSubs);
-    auto input = reqtOrigTy->getInput().subst(reqtSubMap);
+    
+    SmallVector<AnyFunctionType::Param, 4> input;
+    reqtOrigTy->substParams(reqtSubMap, input);
```
Comment (slavapestov) [inline lib/SILGen/SILGenType.cpp]: Why not just reqtOrigTy->substGenericArgs(reqtSubMap)?

File: lib/SILGen/SILGenType.cpp [hunk 3]
```diff
@@ -607,7 +609,9 @@ SILGenModule::emitProtocolWitness(ProtocolConformance *conformance,
         concreteTy,
         ProtocolConformanceRef(specialized));
 
-    auto input = reqtOrigTy->getInput().subst(reqtSubs)->getCanonicalType();
+    SmallVector<AnyFunctionType::Param, 4> input;
+    reqtOrigTy->substParams(reqtSubs, input,
```
Comment (slavapestov) [inline lib/SILGen/SILGenType.cpp]: Ditto

**Review by slavapestov — CHANGES_REQUESTED** [review]

**Comment by CodaFi** [comment]
#10889 and John's changes make this patch pretty useless.

**Pull request closed without merging**